dev/tools/test/localization/gen_l10n_test.dart

```diff
@@ -0,0 +1,627 @@
+          classNameString: defaultClassNameString,
+        );
+        generator.parseArbFiles();
+      } on L10nException catch (e) {
+        fail('Setting language and locales should not fail: \n$e');
+      }
+
+      // @@locale property should hold higher priority
+      expect(generator.supportedLocales.contains(LocaleInfo.fromString('en')), true);
+      expect(generator.supportedLocales.contains(LocaleInfo.fromString('zh')), true);
+      // filename should not be used since @@locale is specified
+      expect(generator.supportedLocales.contains(LocaleInfo.fromString('es')), false);
+      expect(generator.supportedLocales.contains(LocaleInfo.fromString('am')), false);
+    });
+
+    test('throws when arb file\'s locale could not be determined', () {
+      fs.currentDirectory.childDirectory('lib').childDirectory('l10n')
+        ..createSync(recursive: true)
+        ..childFile('app.arb')
+        .writeAsStringSync(singleMessageArbFileString);
+      try {
+        final LocalizationsGenerator generator = LocalizationsGenerator(fs);
+        generator.initialize(
+          l10nDirectoryPath: defaultArbPathString,
+          templateArbFileName: 'app.arb',
+          outputFileString: defaultOutputFileString,
+          classNameString: defaultClassNameString,
+        );
+        generator.parseArbFiles();
+      } on L10nException catch (e) {
+        expect(e.message, contains('locale could not be determined'));
+        return;
+      }
+      fail(
+        'Since locale is not specified, setting languages and locales '
+        'should fail'
+      );
+    });
+    test('throws when the same locale is detected more than once', () {
+      const String secondMessageArbFileString = '''{
+  "market": "MARKET",
+  "@market": {
+    "description": "Label for the Market tab"
+  }
+}''';
+
+      final Directory l10nDirectory = fs.currentDirectory.childDirectory('lib').childDirectory('l10n')
+        ..createSync(recursive: true);
+      l10nDirectory.childFile('app_en.arb')
+        .writeAsStringSync(singleMessageArbFileString);
+      l10nDirectory.childFile('app2_en.arb')
+        .writeAsStringSync(secondMessageArbFileString);
+
+      try {
+        final LocalizationsGenerator generator = LocalizationsGenerator(fs);
+        generator.initialize(
+          l10nDirectoryPath: defaultArbPathString,
+          templateArbFileName: defaultTemplateArbFileName,
+          outputFileString: defaultOutputFileString,
+          classNameString: defaultClassNameString,
+        );
+        generator.parseArbFiles();
+      } on L10nException catch (e) {
+        expect(e.message, contains('Multiple arb files with the same locale detected'));
+        return;
+      }
+
+      fail(
+        'Since en locale is specified twice, setting languages and locales '
+        'should fail'
+      );
+    });
+  });
+
+  group('LocalizationsGenerator.generateClassMethods:', () {
+    test('correctly generates a simple message:', () {
```

Comment:
I think it would be preferable to generate a class and then verify that it's methods DTRT, rather than checking that the code is exactly what we expect. Maybe we can schedule an update to theses tests in a future PR.

Reply:
I agree, I was thinking about this as well, but wasn't sure how to write the test that would test the individual methods the way you're describing. The only way I could think of was generating the class file and then somehow executing a script that makes use of the generated class to verify its behavior